### ISSUE TITLE
Flash erases the recovery mechanism too. 

### ISSUE THREAD

**Issue by hubbsvtgc** [psherman42/Demystifying-OpenOCD#2]
The flash erase command erases from flash base address 0x20000000. 

Would n't this erase the boot-loader (some kind of recovery mechanism) located at 0x20000000 - 0x2000FFFF ?

Reference link:
https://www.reddit.com/r/RISCV/comments/vy6klv/why_hifive1revb_board_boot_process_dont_jump_to/

**Comment by psherman42** [comment]
Yes, that is true. The bootloader would be erased by the flash-write command (targeting 0x2000 0000).

This is a second stage bootloader, though (Zeroth-stage bootloader being what's in the Mask ROM). The second stage bootloader is just a simple application program and nothing serious or special for basic operation of the FE310 SoC.

The "Flash Recovery Mechanism" (FRM), on the other hand, is a tiny bit of code lurking within OTP memory. This FRM starts at 0x2 0010, and it is from there which the jump to 0x2000 0000 (base of SPI Flash) occurs.

See https://forums.sifive.com/t/is-there-c-boot-code-source-for-revb02-hardware/5987/2 for further details.

The FRM was made in a proprietary library and never released. Basically, it uses the BACKUP1 register of the AON block to maintain a persistent count of attempts to successfully query an SPI Flash with its "READ" command, which should be 0x03 for all JEDEC compliant devices. After five unsuccessful attempts, even across reset events and warm starts, the FRM goes into a forever busy wait -- the only way to clear that unfortunate 'forever busy wait' is to power-cycle the CPU (i.e., the board).

Reprogramming OTP for a simple jump to Flash (if it's not 0xFF blank, or a jump to RAM if it's not blank, either) would be the thing to do. However, the OTP programming procedure is a bit touchy and unforgiving.

Of course, none of this is necessary when the MSEL "Boot Mode" pins are made available so that the value of 0x1000 is not always "0b11" (3 decimal, boot-to-spi-flash) at startup
